Document conditions for popup-opening by window.open, and BarProp values

arai-a opened the pull request:
for https://github.com/whatwg/html/pull/6530

 * Added popup feature
 * Added description about requesting popup
 * Changed examples not to use the backward-compat-only UI parts features
 * Updated BarProp.visible to reflect popup request
 * Moved outerWidth/outerHeight to obsolete features page

#### Summary

Update `windowFeatures` parameter behavior for `window.open`, and also `BarProp.visible` behavior, to follow the spec change in https://github.com/whatwg/html/pull/6530.

#### Motivation

Reflect the latest spec.

#### Supporting details

Firefox: fixed in 96 (currently Nightly) https://bugzilla.mozilla.org/show_bug.cgi?id=1701001
Chrome: in progress https://bugs.chromium.org/p/chromium/issues/detail?id=1192701
Safari: bug is filed https://bugs.webkit.org/show_bug.cgi?id=223751

#### Related issues

 * https://github.com/mdn/browser-compat-data/pull/13322

#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [ ] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

arai-a committed:
Standardize the condition for opening a popup by window.open, and BarProp values

for https://github.com/whatwg/html/pull/6530

 * Added popup feature
 * Added description about requesting popup
 * Changed examples not to use the backward-compat-only UI parts features
 * Updated BarProp.visible to reflect popup request
 * Moved outerWidth/outerHeight to obsolete features page

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -50,11 +50,10 @@ var window = window.open(url, windowName, [windowFeatures]);
-    toolbar, and so forth. There must be no whitespace in the string. See
-    {{anch("Window features")}} below for documentation of each of the features that can
-    be specified.
+    given with their corresponding values in the form "name=value", or "name" for boolean feature. These
+    features include options such as the window's default size and position, whether or
+    not to open a minimal-popup window, and so forth. See {{anch("Window features")}}
+    below for documentation of each of the features that can be specified.
```
```suggestion
    given with their corresponding values in the form "name=value", or "name" for boolean features. These
    features include options such as the window's default size and position, whether or
    not to open a minimal-popup window, and so forth. See {{anch("Window features")}}
    below for documentation of each of the features that can be specified.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

arai-a committed:
Update index.md

Remove unnecessary hash for fragment

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,36 @@ parameter is not provided (or if the `windowFeatures` parameter is
-> and platforms (See [popup condition](#popup_condition) section)
+### Popup feature
+
+`windowFeatures` can be used to explicitly request the browser to use a popup window,
+that has minimal UI parts, for the new secondary window.
+
+Whether or not to use a popup window affects {{domxref("BarProp.visible")}} value.
+
+> **Note:** In some browsers, users can configure not to use a popup window.  Also, some
+> browsers, such as mobile browsers, don't have the concept of windows.
+
+The feature can be set to `yes` or `1`, or just be present to be on. Set them to
+`no` or `0`, or in most cases just omit them, to be off.
+
+Example: `popup=yes`, `popup=1`, and `popup` have identical results.
+
+- `popup`
+
+  - : If this feature is present and on, it requests the browser to use a minimal pop-up
+      window for the new secondary window.
+      If this feature is present and off, it requests the browser not to use minimal
+      pop-up window for the secondary window.
```
```suggestion
`windowFeatures` can be used to explicitly request that the browser use a popup window
with minimal UI parts for the new secondary window.

Whether or not to use a popup window affects the {{domxref("BarProp.visible")}} value.

> **Note:** In some browsers, users can configure the browser to not to use a popup window.  Also, some
> browsers, such as mobile browsers, don't have the concept of windows.

To enable the feature, specify `popup` either with no value at all, or else set it to `yes` or `1`.

Example: `popup=yes`, `popup=1`, and `popup` have identical results.

To disable the feature, ether don’t specify it all all (omit it), or else set it to `no` or `0`.

- `popup`

  - : If this feature is present and enabled, it requests that the browser use a minimal pop-up
      window for the new secondary window.
      If this feature is not present, or is present and disabled, it requests that the browser not use a minimal
      pop-up window for the secondary window.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

arai-a commented:
Sorry, my description was unclear.

for "popup" feature, "not-present" and "present with disabled" has different semantics.
"not-present" fallback to other conditions, and "present with disabled" explicitly "don't request" without checking other features.
https://html.spec.whatwg.org/#popup-window-is-requested

sideshowbarker commented:
> Sorry, my description was unclear.
> 
> for "popup" feature, "not-present" and "present with disabled" has different semantics.
> "not-present" fallback to other conditions, and "present with disabled" explicitly "don't request" without checking other features.
> https://html.spec.whatwg.org/#popup-window-is-requested

Ok, thanks ー I'll make another edit to re-correct that

arai-a commented:
the basic algorithm that the document is supposed to cover is that:
  * if "popup" is present, then:
    * if enabled (boolean value is true), then request popup
    * otherwise, do not request popup
  * otherwise:
    * if windowFeatures is empty, then do not request popup
    * otherwise:
      * if there's any other feature than noopener and noreferer, then request popup
        * (this has more complex backward-compat condition about UI features, but I don't want to describe it in the document, for simplicity)
      * otherwise, do not request popup

so I mentioned about popup in position and size (left/top/width/height),
and not mentioned about popup in noopener/noreferer.

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,37 @@ parameter is not provided (or if the `windowFeatures` parameter is
 
-`windowFeatures` parameter can specify the position and size of
-the new window.
+`windowFeatures` parameter can specify the position and size of the new secondary window.
+If any of them are given, and the `popup` feature is not given, it requests the browser
+to use a minimal pop-up window for the secondary window.
+
+> **Note:** In some browsers, users can override the behavior. Also this has no effect on
+> the mobile browsers without the concept of windows.
```
```suggestion
The _windowFeatures_ parameter can specify the position and size of the new secondary window,
by specifying feature names either using _name=value_ pairs, or for boolean features, using just a _name_.
If any feature names are given, and the `popup` feature name is not given, it requests that the browser
use a minimal pop-up window for the secondary window.

> **Note:** In some browsers, users can override this behavior. Also this also has no effect in
> mobile browsers, which lack the concept of windows.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/barprop/visible/index.md:
```diff
@@ -21,11 +21,16 @@ let visible = BarProp.visible;
 ### Value
 
-A {{jsxref("Boolean")}}, which is true if the bar represented by the used interface element is visible.
+A {{jsxref("Boolean")}}, which is true if the top-level window is opened by
+{{domxref("window.open")}} with {{domxref("window.open", "requesting a popup window", "popup_feature", 1)}}.
+
+> **Note:** Historically this represented whether the used interface element is visible
+> or not. For privacy reasons, this no more represent the actual visibility of each
+> interface element.
```
```suggestion
A {{jsxref("Boolean")}}, which is true if the top-level window is opened by
{{domxref("window.open")}} with the {{domxref("window.open", "requesting a popup window", "popup_feature", 1)}}.

> **Note:** Historically this represented whether the interface element used is visible
> or not. But for privacy reasons, this no longer represents the actual visibility of each
> interface element.
```

sideshowbarker committed:
Update files/en-us/web/api/barprop/visible/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -50,11 +50,10 @@ var window = window.open(url, windowName, [windowFeatures]);
-    toolbar, and so forth. There must be no whitespace in the string. See
-    {{anch("Window features")}} below for documentation of each of the features that can
-    be specified.
+    given with their corresponding values in the form "name=value", or "name" for boolean features. These
+    features include options such as the window's default size and position, whether or
+    not to open a minimal-popup window, and so forth. See {{anch("Window features")}}
+    below for documentation of each of the features that can be specified.
```
```suggestion
    given with their corresponding values in the form _name=value_ — or for boolean features, just _name_. These
    features include options such as the window's default size and position, whether or
    not to open a minimal-popup window, and so forth. See {{anch("Window features")}}
    below for documentation of each of the features that can be specified.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,38 @@ parameter is not provided (or if the `windowFeatures` parameter is
-> and platforms (See [popup condition](#popup_condition) section)
+### Popup feature
+
+`windowFeatures` can be used to explicitly request that the browser use a popup window
+with minimal UI parts for the new secondary window.
```
```suggestion
The _windowFeatures_ features parameter can be used to explicitly request that the browser use a popup window
with minimal UI parts for the new secondary window.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,38 @@ parameter is not provided (or if the `windowFeatures` parameter is
+
+Whether or not to use a popup window affects the {{domxref("BarProp.visible")}} value.
+
+> **Note:** In some browsers, users can configure the browser to not to use a popup window.  Also, some
+> browsers, such as mobile browsers, don't have the concept of windows.
```
```suggestion
> **Note:** In some browsers, users can configure the browser to not to use a popup window. Also, some
> browsers, such as mobile browsers, don't have the concept of windows.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,38 @@ parameter is not provided (or if the `windowFeatures` parameter is
+If any feature names are given, and the `popup` feature name is not given, it requests that the browser
+use a minimal pop-up window for the secondary window.
+
+> **Note:** In some browsers, users can override this behavior. Also this also has no effect in
+> mobile browsers, which lack the concept of windows.
```
```suggestion
> **Note:** In some browsers, users can override this behavior. Also, it has no effect in
> mobile browsers, which lack the concept of windows.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -50,11 +50,10 @@ var window = window.open(url, windowName, [windowFeatures]);
-    toolbar, and so forth. There must be no whitespace in the string. See
-    {{anch("Window features")}} below for documentation of each of the features that can
-    be specified.
+    given with their corresponding values in the form _name=value_ — or for boolean features, just _name_. These
+    features include options such as the window's default size and position, whether or
+    not to open a minimal-popup window, and so forth. See {{anch("Window features")}}
+    below for documentation of each of the features that can be specified.
```
```suggestion
    given with their corresponding values in the form _name=value_ — or for boolean features, just _name_. These
    features include options such as the window's default size and position, whether or
    not to open a minimal popup window, and so forth. See {{anch("Window features")}}
    below for documentation of each of the features that can be specified.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,38 @@ parameter is not provided (or if the `windowFeatures` parameter is
+
+Example: `popup=yes`, `popup=1`, and `popup` have identical results.
+
+To disable the feature, ether don’t specify it all all (omit it), or else set it to `no` or `0`.
```
```suggestion
Otherwise:

* To not request a popup, omit the _windowFeatures_ parameter.
* Otherwise:
   * Specifying any features other than `noopener` or `noreferer` has the effect of also requesting a popup.
   * Otherwise, no popup is requested.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,43 @@ parameter is not provided (or if the `windowFeatures` parameter is
+> **Note:** In some browsers, users can configure the browser to not to use a popup window. Also, some
+> browsers, such as mobile browsers, don't have the concept of windows.
+
+To enable the feature, specify `popup` either with no value at all, or else set it to `yes` or `1`.
+
+Example: `popup=yes`, `popup=1`, and `popup` have identical results.
+
+Otherwise:
+
+* To not request a popup, omit the _windowFeatures_ parameter.
+* Otherwise:
+   * Specifying any features other than `noopener` or `noreferer` has the effect of also requesting a popup.
+   * Otherwise, no popup is requested.
+
```
```suggestion
```

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,43 @@ parameter is not provided (or if the `windowFeatures` parameter is
+      window for the new secondary window.
+      If this feature is not present, or is present and disabled, it requests that the browser not use a minimal
+      pop-up window for the secondary window.
 
```
```suggestion

To enable the feature, specify `popup` either with no value at all, or else set it to `yes` or `1`.

Example: `popup=yes`, `popup=1`, and `popup` all have identical results.

Otherwise:

* To not request a popup, omit the _windowFeatures_ parameter.
* Otherwise:
   * Specifying any features in the _windowFeatures_ parameter other than `noopener` or `noreferer` has the effect of also requesting a popup.
   * Otherwise, no popup is requested.

```

sideshowbarker committed:
Apply suggestions from code review

sideshowbarker commented:
> Sorry, my description was unclear.
> 
> for "popup" feature, "not-present" and "present with disabled" has different semantics. "not-present" fallback to other conditions, and "present with disabled" explicitly "don't request" without checking other features. [html.spec.whatwg.org/#popup-window-is-requested](https://html.spec.whatwg.org/#popup-window-is-requested)
> …
> the basic algorithm that the document is supposed to cover is that:
> 
>     * if "popup" is present, then:
>       
>       * if enabled (boolean value is true), then request popup
>       * otherwise, do not request popup
> 
>     * otherwise:
>       
>       * if windowFeatures is empty, then do not request popup
>       * otherwise:
>         
>         * if there's any other feature than noopener and noreferer, then request popup
>           
>           * (this has more complex backward-compat condition about UI features, but I don't want to describe it in the document, for simplicity)
>         * otherwise, do not request popup

OK, please see my latest edit, and let me know if it looks accurate now

arai-a committed:
Update index.md

the "not present" case is covered by "Otherwise" -> "Otherwise, no popup is requested.", that's after "Specifying any features in the _windowFeatures_ parameter other than `noopener` or `noreferer` has the effect of also requesting a popup." part.
so I think it's better omitting the "not present" case from the first section.

arai-a commented:
oh, I thought the description becomes comment instead of commit message.

anyway, thank you for reviewing :)
I think it's better omitting the "not present" case from the first section, and let the later "Otherwise" handle it,
so that it checks other features.

arai-a commented:
the difference is that:
`width=500,height=500` (not present) requests popup, and
`width=500,height=500,popup=no` (present and disabled) doesn't request popup.

(in both case, whether to handle width/height, and whether to open new window or tab, are browser-dependent)

sideshowbarker commented:
> oh, I thought the description becomes comment instead of commit message.
> 
> anyway, thank you for reviewing :) I think it's better omitting the "not present" case from the first section, and let the later "Otherwise" handle it, so that it checks other features.

Please go ahead and commit a change for that

arai-a commented:
done in https://github.com/mdn/content/pull/10339/commits/dbd3667a0ac0707a9c0c4ade3e1fdd036b7a174d

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,43 @@ parameter is not provided (or if the `windowFeatures` parameter is
 
-`windowFeatures` parameter can specify the position and size of
-the new window.
+The _windowFeatures_ parameter can specify the position and size of the new secondary window,
+by specifying feature names either using _name=value_ pairs, or for boolean features, using just a _name_.
+If any feature names are given, and the `popup` feature name is not given, it requests that the browser
+use a minimal pop-up window for the secondary window.
```
```suggestion
You can use the _windowFeatures_ parameter to control the position and size of the new secondary window,
by specifying feature names either using _name=value_ pairs, or for boolean features, using just a _name_.
If any feature names are given, and the `popup` feature name is not given, it requests that the browser
use a minimal pop-up window for the secondary window.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -178,31 +205,15 @@ If the `windowFeatures` parameter is non-empty and no size
-> **Warning:** In modern browsers (Firefox 76 or newer, Google Chrome, Safari, Chromium Edge), the
+> **Warning:** These features are kept only for backward compatibility.
+> In modern browsers (Firefox 76 or newer, Google Chrome, Safari, Chromium Edge), the
 > following features are just a condition for whether to open popup or not. See [popup condition](#popup_condition) section.
```
```suggestion
> following features are just a condition for whether to open popup or not. See the [popup condition](#popup_condition) section.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -178,31 +205,15 @@ If the `windowFeatures` parameter is non-empty and no size
-`no` or `0`, or in most cases just omit them, to be off.
+> **Warning:** These features are kept only for backward compatibility.
+> In modern browsers (Firefox 76 or newer, Google Chrome, Safari, Chromium Edge), the
+> following features are just a condition for whether to open popup or not. See the [popup condition](#popup_condition) section.
```
```suggestion
> following features are just a condition for whether to open a popup or not. See the [popup condition](#popup_condition) section.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -178,31 +205,15 @@ If the `windowFeatures` parameter is non-empty and no size
 
-Example: `status=yes`, `status=1`, and `status` have
-identical results.
+The following features control the visibility of each UI parts, these features can also
+be set to `yes` or `1`, or just be present to be on. Set them to `no` or `0`, or in most
+cases just omit them, to be off.
```
```suggestion
The following features control the visibility of each UI part.
To enable them, either specify them with no value at all, or else set them to `yes` or `1`.
To disable them, either omit them, or else set them to `no` or `0`.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -671,7 +681,7 @@ UI-related items of `windowFeatures` are used as a condition to
 of them is fixed.
 
-The condition is implementation-dependent and not guaranteed to be stable.
+The detailed condition is described in ["To check if a popup window is requested"](https://html.spec.whatwg.org/#popup-window-is-requested) section in the spec.
```
```suggestion
The detailed condition is described in detail in the ["To check if a popup window is requested"](https://html.spec.whatwg.org/multipage/window-object.html#popup-window-is-requested) section in the spec.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/obsolete_features/index.md:
```diff
@@ -13,3 +13,13 @@ This page lists the obsolete `windowFeatures` parameter of [window.open](/en-US/
 - `personalbar` {{deprecated_inline}}
   - : If this feature is on, then the new secondary window renders the Personal Toolbar in Netscape 6.x, Netscape 7.x and Mozilla browser. It renders the Bookmarks Toolbar in Firefox. In addition to the Personal Toolbar, Mozilla browser will render the Site Navigation Bar if such toolbar is visible, present in the parent window.
     Mozilla and Firefox users can force new windows to always render the Personal Toolbar/Bookmarks toolbar by setting `dom.disable_window_open_feature.personalbar` to _true_ in [about:config](http://support.mozilla.com/en-US/kb/Editing+configuration+files#about_config) or in their [user.js file](http://support.mozilla.com/en-US/kb/Editing+configuration+files#user_js).
+- `outerWidth` {{deprecated_inline}} (only on Firefox, obsolete from Firefox 80)
+  - : Specifies the width of the whole browser window in pixels. This
+    `outerWidth` value includes the window vertical scrollbar (if present) and
+    left and right window resizing borders.
+- `outerHeight` {{deprecated_inline}} (only on Firefox, obsolete from Firefox 80)
+  - : Specifies the height of the whole browser window in pixels. This
+    `outerHeight` value includes any/all present toolbar, window horizontal
+    scrollbar (if present) and top and bottom window resizing borders. Minimal required
+    value is 100.
```
```suggestion
- `outerWidth` {{deprecated_inline}} (only in Firefox; obsolete from Firefox 80)
  - : Specifies the width of the whole browser window, in pixels. This
    `outerWidth` value includes the window vertical scrollbar (if present) and
    left and right window resizing borders.
- `outerHeight` {{deprecated_inline}} (only in Firefox; obsolete from Firefox 80)
  - : Specifies the height of the whole browser window, in pixels. This
    `outerHeight` value includes any and all present toolbars, the window horizontal
    scrollbar (if present), and top and bottom window-resizing borders. The minimal required
    value is 100.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/obsolete_features/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,43 @@ parameter is not provided (or if the `windowFeatures` parameter is
-> and platforms (See [popup condition](#popup_condition) section)
+### Popup feature
+
+The _windowFeatures_ features parameter can be used to explicitly request that the browser use a popup window
```
```suggestion
The _windowFeatures_ parameter can be used to explicitly request that the browser use a popup window
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -121,15 +120,43 @@ parameter is not provided (or if the `windowFeatures` parameter is
-> and platforms (See [popup condition](#popup_condition) section)
+### Popup feature
+
+The _windowFeatures_ parameter can be used to explicitly request that the browser use a popup window
+with minimal UI parts for the new secondary window.
```
```suggestion
You can specify the _windowFeatures_ parameter to explicitly request that the browser use a popup window
with minimal UI parts for the new secondary window.
```

sideshowbarker committed:
Update files/en-us/web/api/window/open/index.md

sideshowbarker commented:
@arai-a Please re-review the current state of the content in the branch. If it all looks accurate to you, then I will merge it later this afternoon. (Right now I’m going to take a break for a couple hours or so.)

arai-a committed:
Remove boolean feature description from position and size

position and size are int value

arai-a commented:
Thank you again!

tweaked the position/size description not to mention boolean feature handling.
otherwise it looks good :)

sideshowbarker commented:
Sorry! Only just now realized I’d forgotten to merge this…

sideshowbarker commented:
@arai-a Really great work (including the related spec work you did) — congrats on landing your first docs change here, and welcome aboard 🎉
